libc/stdlib_alloc.c

```diff
@@ -0,0 +1,561 @@
+#pragma GCC diagnostic ignored "-Walloc-size-larger-than="
+
+	errno = 0;
+	TEST_ASSERT_NULL(calloc(1, SIZE_MAX));
+	TEST_ASSERT_EQUAL_INT(ENOMEM, errno);
+
+	errno = 0;
+	TEST_ASSERT_NULL(calloc(SIZE_MAX, 1));
+	TEST_ASSERT_EQUAL_INT(ENOMEM, errno);
+
+	errno = 0;
+	TEST_ASSERT_NULL(calloc(SIZE_MAX, SIZE_MAX));
+	TEST_ASSERT_EQUAL_INT(ENOMEM, errno);
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(stdlib_malloc, realloc_null)
+{
+	int *ptr = realloc(NULL, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	int *ptr1 = realloc(NULL, 0);
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	free(ptr);
+	free(ptr1);
+}
+
+
+TEST(stdlib_malloc, realloc_zero_size)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, *ptr);
+#ifndef __phoenix__
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr));
+#endif
+
+	int *ptr1 = realloc(ptr, 0);
+	TEST_ASSERT_NULL(ptr1);
+}
+
+
+TEST(stdlib_malloc, realloc_calloc_resize)
+{
+	int NUMBERS[BLOCK_SIZE];
+	int *ptr = calloc(BLOCK_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE);
+	for (int i = 0; i < BLOCK_SIZE; i++) {
+		ptr[i] = i;
+		NUMBERS[i] = i;
+	}
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr, BLOCK_SIZE);
+	TEST_ASSERT_EQUAL_INT(0, (uintptr_t)ptr % sizeof(int));
+
+	int *ptr1 = realloc(ptr, BLOCK_SIZE * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr1, BLOCK_SIZE);
+	TEST_ASSERT_EQUAL_INT(0, (uintptr_t)ptr1 % sizeof(int));
+
+	free(ptr1);
+}
+
+
+TEST(stdlib_malloc, realloc_calloc_resize_smaller)
+{
+	int *ptr = calloc(1, 2 * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, ptr[0]);
+	ptr[0] = 1;
+
+	int *ptr1 = realloc(ptr, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT(1, ptr1[0]);
+	TEST_ASSERT_EQUAL_INT(0, (uintptr_t)ptr1 % sizeof(int));
+
+	int *ptr2 = calloc(1, 2 * (BLOCK_SIZE * LARGE_SIZE));
+	TEST_ASSERT_NOT_NULL(ptr2);
+	TEST_ASSERT_EQUAL_INT(0, ptr2[0]);
+	ptr2[0] = 1;
+
+	int *ptr3 = realloc(ptr2, BLOCK_SIZE * LARGE_SIZE);
+	TEST_ASSERT_NOT_NULL(ptr3);
+	TEST_ASSERT_EQUAL_INT(1, ptr3[0]);
+
+	free(ptr1);
+	free(ptr3);
+}
+
+
+TEST(stdlib_malloc, realloc_calloc_resize_larger)
+{
+	int NUMBERS[BLOCK_SIZE];
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, *ptr);
+	ptr[0] = 1;
+
+	int *ptr1 = realloc(ptr, 2 * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT(0, (uintptr_t)ptr1 % sizeof(int));
+	TEST_ASSERT_EQUAL_INT(1, ptr1[0]);
+
+	int *ptr2 = calloc(BLOCK_SIZE / 8, BLOCK_SIZE * LARGE_SIZE);
+	TEST_ASSERT_NOT_NULL(ptr2);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr2, BLOCK_SIZE);
+	for (int i = 0; i < BLOCK_SIZE; i++) {
+		ptr2[i] = i;
+		NUMBERS[i] = i;
+	}
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr2, BLOCK_SIZE);
+
+	int *ptr3 = realloc(ptr2, 2 * (BLOCK_SIZE * LARGE_SIZE));
+	TEST_ASSERT_NOT_NULL(ptr3);
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr3, BLOCK_SIZE);
```

Comment:
```suggestion
	for (int i = 0; i < BLOCK_SIZE * LARGE_SIZE; i++) {
	    TEST_ASSERT_EQ(ptr[i], i)
	}
```
If you want to use `unsigned char*` type make sure to take modulo: `TEST_ASSERT_EQ(ptr[i], i % sizeof(unsigned char))`